### ISSUE TITLE
Misleading error messages (due to missing analysis of control flow?)

### ISSUE THREAD

**Issue by tos-kamiya** [microsoft/pyright#3821]
Misleading error message when member variables are assigned in multiple places.

To reproduce, create the following test script `test_w_unittest.py` and apply pyright to it: `pyright test_w_unittest.py`.

* Two errors were reported in a class TestConfusingErrorReported.
* No errors were reported in a class TestErrorMissed.

```python
import unittest
import tempfile


class TestOK(unittest.TestCase):
    def setUp(self):
        self.temp_dir = tempfile.TemporaryDirectory()

    def tearDown(self):
        self.temp_dir.cleanup()

    def test_bar(self):
        print("temp dir name: %s" % self.temp_dir.name)


class TestConfusingErrorReported(unittest.TestCase):
    def setUp(self):
        self.temp_dir = None
        self.temp_dir = tempfile.TemporaryDirectory()

    def tearDown(self):
        self.temp_dir.cleanup()  # error: "cleanup" is not a known member of "None" (reportOptionalMemberAccess)

    def test_bar(self):
        print("temp dir name: %s" % self.temp_dir.name)  # error: "cleanup" is not a known member of "None" (reportOptionalMemberAccess)


class TestErrorMissed(unittest.TestCase):
    def setUp(self):
        self.temp_dir.cleanup()

    def tearDown(self):
        self.temp_dir = tempfile.TemporaryDirectory()

    def test_bar(self):
        print("temp dir name: %s" % self.temp_dir.name)
```

Screenshot:

![Screenshot from 2022-08-12 16-50-31](https://user-images.githubusercontent.com/1262286/184309813-62450e7d-52a2-43a5-94ed-c4fd78875cbb.png)

To my understanding, the best output is:
* No error found in TestConfusingErrorReported 
* Errors found in TestErrorMissed

The pyright version was 1.1.266 as shown in the screenshot.

Do I misunderstand something?

**Comment by erictraut** [comment]
This is working correctly. The type of instance variable `temp_dir` is not declared (i.e. it has no annotation), so its type is inferred based on all assigned values within the class implementation. In your sample, it is assigned values of `None` and an instance of `TemporaryDirectory[str]`, so its type is inferred to be a union of these two types.

If your intent is for the variable to contain only a `TemporaryDirectory[str]` instance, you can declare it as such or avoid assigning a value of a different type.

Pyright does perform code flow analysis locally for type narrowing purposes, but it does not perform code flow analysis globally. In general, a static analyzer cannot determine which order methods will be called in.

**Comment by tos-kamiya** [comment]
The conclusion is that I should not rely on global control flow analysis but should explicitly add type hints, as follows:
`self.temp_dir: tempfile.TemporaryDirectory[str] = tempfile.TemporaryDirectory()`

I now have a clear image of how to use pyright.
Thank you very much.
